fe/fe-core/src/main/java/com/starrocks/connector/paimon/PaimonMetadata.java

```diff
@@ -292,4 +300,79 @@ public long getTableCreateTime(String dbName, String tblName) throws Exception {
         }
         return 0;
     }
+
+    public List<PaimonPartitionInfo> getChangedPartitionInfo(PaimonTable table, long latestSnapShotID) {
+        List<PaimonPartitionInfo> result = new ArrayList<>();
+        Map<String, Long> partitionToSnapshotId = monitor(table, latestSnapShotID);
+        for (Map.Entry<String, Long> entry : partitionToSnapshotId.entrySet()) {
+            PaimonPartitionInfo partitionInfo = new PaimonPartitionInfo(entry.getKey(), entry.getValue());
+            result.add(partitionInfo);
+            if (entry.getValue() != null) {
+                latestSnapShotID = Math.max(latestSnapShotID, entry.getValue());
+            }
+        }
+        table.setLastedSnapshotId(latestSnapShotID);
+        LOG.debug("Get changed partitionInfo:{}", result);
+        return result;
+    }
+
+    private Map<String, Long> monitor(PaimonTable paimonTable, long nextSnapshotId) {
+        Map<String, Long> partitionToSnapshotId = new HashMap<>();
+        FileMonitorTable fileMonitorTable = new FileMonitorTable(paimonTable.getNativeTable());
+        Long latestId = fileMonitorTable.snapshotManager().latestSnapshotId();
+        long latestSnapshotId = latestId == null ? Long.MIN_VALUE : latestId;
+        ReadBuilder readBuilder = fileMonitorTable.newReadBuilder();
+        StreamTableScan scan = readBuilder.newStreamScan();
+        TableRead read = readBuilder.newRead();
+        if (nextSnapshotId != Long.MIN_VALUE) {
+            scan.restore(nextSnapshotId);
+        } else {
+            scan.restore(null);
+        }
+        try {
+            while (true) {
+                if (!hashNext(paimonTable, scan, read, partitionToSnapshotId)) {
+                    break;
+                }
+            }
+        } catch (OutOfRangeException e) {
+            partitionToSnapshotId.clear();
+            List<String> parts = listPartitionNames(paimonTable.getDbName(), paimonTable.getTableName());
+            for (String part : parts) {
+                partitionToSnapshotId.put(part, latestSnapshotId);
+            }
+        }
+        return partitionToSnapshotId;
+    }
+
+    private boolean hashNext(PaimonTable paimonTable, StreamTableScan scan, TableRead read,
+                             Map<String, Long> partitionToSnapshotId) {
+        TableScan.Plan plan = scan.plan();
+        if (plan.splits().isEmpty()) {
+            return false;
+        }
+        try {
+            read.createReader(plan).forEachRemaining(new Consumer<InternalRow>() {
```

Comment:
why we need create reader every time when we call this api

Reply:
Because reader is create only for once plan. Every plan need a new reader.